presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -1313,7 +1369,13 @@ public HiveInsertTableHandle beginInsert(ConnectorSession session, ConnectorTabl
                 .collect(toList());
 
         HiveStorageFormat tableStorageFormat = extractHiveStorageFormat(table.get());
-        LocationHandle locationHandle = locationService.forExistingTable(metastore, session, table.get());
+        LocationHandle locationHandle;
+        if (table.get().getTableType().equals(TEMPORARY_TABLE)) {
+            locationHandle = locationService.forTemporaryTable(metastore, session, tableName.getSchemaName(), tableName.getTableName());
```

Comment:
Note `forTemporaryTable` get called twice: once in `createTemporaryTable` , once in `beginInsert`. 

The current implementation  relies on the fact that `LocationService` will generate same target directory, given the same table name, which seems to be a bit too strong. 


Maybe make `forTemporaryTable` to accept an optional table parameter, and use the location specified in `table.getStorage()` if presents?  What do you think?

Reply:
Let's just have 2 separate methods. One that accepts `schemaName, tableName`, and one that accepts `table`.

Reply:
Now it is called only once. Please have a look at the set of changes that were made: https://github.com/prestodb/presto/compare/1769c0a2b08d7a4c524e7632ffdac58796b0f034..1905f27ba5a521761ffcd60613b36df47b7eb82e